File: sdks/ts/memory/src/ratelimit/factory.ts
// SPDX-License-Identifier: Apache-2.0

/**
 * Per-tenant rate limiter factory. Each tenant receives an isolated
 * token bucket (optionally wrapped in an adaptive layer). Idle tenants
 * are evicted after a configurable TTL to prevent unbounded growth.
 */

import { noopLogger } from '../llm/types.js'
import { createAdaptiveRateLimiter } from './adaptive.js'
import { createTokenBucket } from './token-bucket.js'
import type { RateLimiter, RateLimiterFactory, RateLimiterFactoryOptions } from './types.js'

/** Default idle TTL: 5 minutes. */
const DEFAULT_TENANT_TTL_MS = 300_000

/** Eviction sweep interval: 60 seconds. */
const EVICTION_INTERVAL_MS = 60_000

type TenantEntry = {
  limiter: RateLimiter
  lastAccessedAt: number
}

/**
 * Create a factory that produces per-tenant rate limiters. Calling
 * `forTenant` with the same ID returns the same limiter instance.
 * Idle tenants are evicted after `tenantTtlMs` milliseconds.
 */
export const createRateLimiterFactory = (opts: RateLimiterFactoryOptions): RateLimiterFactory => {
  if (opts.defaultMaxTokens <= 0) {
    throw new Error('defaultMaxTokens must be a positive number')
  }
  if (opts.defaultRefillRate <= 0) {
    throw new Error('defaultRefillRate must be a positive number')
  }

  const logger = opts.logger ?? noopLogger
  const tenantTtlMs = opts.tenantTtlMs ?? DEFAULT_TENANT_TTL_MS
  const tenants = new Map<string, TenantEntry>()
  let closed = false

  const evictStale = (): void => {
    const now = Date.now()
    for (const [id, entry] of tenants) {
      if (now - entry.lastAccessedAt > tenantTtlMs) {
        tenants.delete(id)
        entry.limiter.close().catch(() => {
          // Swallow close errors during eviction.
        })
        logger.info('evicted idle tenant rate limiter', { tenantId: id })
      }
    }
  }

  // Periodic eviction sweep.
  const evictionTimer = setInterval(evictStale, EVICTION_INTERVAL_MS)
  if (typeof evictionTimer === 'object' && 'unref' in evictionTimer) {
    evictionTimer.unref()
  }

  const forTenant = (tenantId: string): RateLimiter => {
    if (closed) {
      throw new Error('rate limiter factory is closed')
    }

    const existing = tenants.get(tenantId)
    if (existing !== undefined) {
      existing.lastAccessedAt = Date.now()
      return existing.limiter
    }

    const bucket = createTokenBucket({
      maxTokens: opts.defaultMaxTokens,
      refillRatePerSecond: opts.defaultRefillRate,
      ...(opts.defaultMaxConcurrency !== undefined ? { maxConcurrency: opts.defaultMaxConcurrency } : {}),
      tenantId,
      logger,
    })

    const limiter: RateLimiter = opts.adaptiveEnabled === true
      ? createAdaptiveRateLimiter({
          bucket,
          ...(opts.minRefillRate !== undefined ? { minRefillRate: opts.minRefillRate } : {}),
          ...(opts.maxRefillRate !== undefined ? { maxRefillRate: opts.maxRefillRate } : {}),
          ...(opts.recoveryFactor !== undefined ? { recoveryFactor: opts.recoveryFactor } : {}),
          logger,
        })
      : bucket

    tenants.set(tenantId, { limiter, lastAccessedAt: Date.now() })
    return limiter
  }

  const close = async (): Promise<void> => {
    closed = true
    clearInterval(evictionTimer)
    const closers = [...tenants.values()].map(entry => entry.limiter.close())
    tenants.clear()
    await Promise.all(closers)
  }

  return { forTenant, close }
}

File: go/ratelimit/factory.go
// SPDX-License-Identifier: Apache-2.0

package ratelimit

import (
	"fmt"
	"log/slog"
	"sync"
	"time"
)

// defaultTenantTTL is the idle duration after which a tenant limiter
// is evicted from the factory. Default: 5 minutes.
const defaultTenantTTL = 5 * time.Minute

// evictionInterval is how often the factory sweeps for stale tenants.
const evictionInterval = 60 * time.Second

// tenantEntry holds a limiter and its last access timestamp.
type tenantEntry struct {
	limiter      Limiter
	lastAccessed time.Time
}

// tenantFactory implements [Factory] with per-tenant limiter isolation.
// Each tenant gets its own token bucket (and optional adaptive wrapper).
// Idle tenants are evicted after TenantTTL.
type tenantFactory struct {
	mu      sync.Mutex
	tenants map[string]*tenantEntry
	opts    FactoryOptions
	logger  *slog.Logger
	closed  bool
	stopCh  chan struct{}
}

// NewFactory creates a [Factory] that produces per-tenant limiters.
// Panics if DefaultMaxTokens or DefaultRefillRate are not positive.
func NewFactory(opts FactoryOptions) Factory {
	if opts.DefaultMaxTokens <= 0 {
		panic(fmt.Sprintf("ratelimit: DefaultMaxTokens must be positive, got %d", opts.DefaultMaxTokens))
	}
	if opts.DefaultRefillRate <= 0 {
		panic(fmt.Sprintf("ratelimit: DefaultRefillRate must be positive, got %f", opts.DefaultRefillRate))
	}

	logger := opts.Logger
	if logger == nil {
		logger = slog.New(slog.DiscardHandler)
	}

	ttl := opts.TenantTTL
	if ttl <= 0 {
		ttl = defaultTenantTTL
	}
	opts.TenantTTL = ttl

	f := &tenantFactory{
		tenants: make(map[string]*tenantEntry),
		opts:    opts,
		logger:  logger,
		stopCh:  make(chan struct{}),
	}

	go f.evictionLoop()
	return f
}

// evictionLoop periodically removes idle tenants.
func (f *tenantFactory) evictionLoop() {
	ticker := time.NewTicker(evictionInterval)
	defer ticker.Stop()

	for {
		select {
		case <-f.stopCh:
			return
		case <-ticker.C:
			f.evictStale()
		}
	}
}

// evictStale removes tenants that have not been accessed within the TTL.
func (f *tenantFactory) evictStale() {
	f.mu.Lock()
	defer f.mu.Unlock()

	now := time.Now()
	for id, entry := range f.tenants {
		if now.Sub(entry.lastAccessed) > f.opts.TenantTTL {
			_ = entry.limiter.Close()
			delete(f.tenants, id)
			f.logger.Info("evicted idle tenant rate limiter", "tenantId", id)
		}
	}
}

func (f *tenantFactory) ForTenant(tenantID string) Limiter {
	f.mu.Lock()
	defer f.mu.Unlock()

	if f.closed {
		panic("ratelimit: ForTenant called on closed factory")
	}

	if entry, ok := f.tenants[tenantID]; ok {
		entry.lastAccessed = time.Now()
		return entry.limiter
	}

	bucket := NewBucket(BucketOptions{
		MaxTokens:        f.opts.DefaultMaxTokens,
		RefillRatePerSec: f.opts.DefaultRefillRate,
		MaxConcurrency:   f.opts.DefaultMaxConcurrency,
		TenantID:         tenantID,
		Logger:           f.logger,
	})

	var lim Limiter = bucket
	if f.opts.AdaptiveEnabled {
		lim = NewAdaptive(AdaptiveOptions{
			Bucket:         bucket,
			MinRefillRate:  f.opts.MinRefillRate,
			MaxRefillRate:  f.opts.MaxRefillRate,
			RecoveryFactor: f.opts.RecoveryFactor,
			Logger:         f.logger,
		})
	}

	f.tenants[tenantID] = &tenantEntry{
		limiter:      lim,
		lastAccessed: time.Now(),
	}
	return lim
}

func (f *tenantFactory) Close() error {
	f.mu.Lock()
	defer f.mu.Unlock()

	f.closed = true
	close(f.stopCh)

	var firstErr error
	for id, entry := range f.tenants {
		if err := entry.limiter.Close(); err != nil && firstErr == nil {
			firstErr = err
		}
		delete(f.tenants, id)
	}
	return firstErr
}
